blog/2019-09-ngrx-data-views/README.md

```diff
@@ -0,0 +1,419 @@
+
+If we check our example selectors we can see how we are doing:
+
+- `getBook`:
+  
+  The projector of this selector references a single value of the entities state of `@ngrx/entity`. Updates to the state are handled by entity adapters and the object referenced will only change if the book with the same ID selected is updated or deleted. Other book updates will modify a different reference, so we are safe here. The third check in `defaultMemoize` will detect object identity for updates to other objects and thus the selector is properly memoized.
+
+- `getAuthorsOfBook`:
+
+  The fatal flaw lies within the projector of this selector:
+
+  ```ts
+  book ? book.authorIds.map(authorId => authors[authorId])` : []
+  ```
+  
+  This is a function that will always create a new array upon every call. So the third check for output identity will most certainly always fail and the selector will fire every time its input changes. In this case every change to the authors state - be it related to the currently selected book or not - will trigger a change.
+
+- the same applies to the `getTagsOfBook` selector: every change to the tags state will trigger view update.
+
+- Our data view selector `getBookView` consumes all these changes and it also has a projector function that creates a new object every time it is called. The issue hereby multiplies as all dependant selectors trigger changes.
+
+Another thing we didn't pay attention to is the fact that our parameterized selectors share a memoization cache if they are statically instantiated. If we use them multiple times on the same page or chain them in other selectors they will eventually always fire because the same selector is invoked with different parameters all the time. We would have to find a way to separate the memoization caches.
+
+All in all, our resulting data view is not properly memoized and hence quite useless.
+
+## Tackling Memoization
+
+What we need to do now to fix this is to assure that `getBookView` returns a correctly memoized result. We can do this directly by adding memoization to `getBookView`, but better yet, we can just assure that the inputs of that selector only trigger a change, when they really change. Our goal must be to fix `getAuthorsOfBook` and `getTagsOfBook`. So let's have another look at `defaultMemoize` first:
+
+```ts
+export function defaultMemoize(
+  projectionFn: AnyFn,
+  isArgumentsEqual = isEqualCheck,
+  isResultEqual = isEqualCheck
+): MemoizedProjection
+```
+
+What we can do is setting up the memoization ourselves. We will now use `createSelectorFactory` for creating the selector directly and reference `defaultMemoize` applied with some overrides. Additionally we do not create a static instance but instead supply a factory method for our parameterized selectors so each instance has its own memoization cache:
+
```

Comment:
Ugh I needed a few attempts to fully understand this. I think this needs a bit more introduction and explanation. the complex code example maybe could be separated by concepts? show the general schema first before applying it to our specific selector

Reply:
Yes, I added this in a hurry. With the last comment, I'll restructure this part. Thanks for the input :slightly_smiling_face: